### PR TITLE
Fighters3

### PR DESCRIPTION
**Pull request by geoffthemedio**
Adds fighters to combat, and three part types to support them. Fighters are launched during battles and attack on subsequent rounds. Any attack will destroy a fighter, regardless of weapon strength. Fighter weapon strength is determined by the best fighter weapon on a ship, and may be zero to have fighters just be decoys in battle. Number of fighters launched per combat round is determined by fighter bay capacity. Total number of fighters that can be launched depends on hangar capacity. Surviving fighters are recovered at end of battles, destroyed fighters are replaced by resupply of their ship. Requires additional parts and extensive balancing and AI adjustments, but is a hopefully distinctive mechanic for weapons.

### CODE REVIEW

**Comment by MatGB** [comment]
[konsole-fighters-compile-error.txt](https://github.com/freeorion/freeorion/files/75832/konsole-fighters-compile-error.txt)

Complete compile failure. Idea looks cool though.

**Comment by geoffthemedio** [comment]
try again now?

**Comment by MatGB** [comment]
Yup, that now compiles and I've got the parts in my design screen. My brain isn't telling me what they do so I'll give it another look in the morning.

**Comment by geoffthemedio** [comment]
You get a demo carrier to play with in your starting battle fleet.

**Comment by MatGB** [comment]
Ah, so you do, yes. OK, first combat, it died turn one against some natives, I think needs to be a Large Hull for demonstration purposes so it's got enough structure to survive a bit.

Question, at strength 2 the fighters are useless against shields unless I'm reading the code wrong. Is it the intent to have them bypass shields or are we looking at tech/species trait increases on them? Because if they don't bypass shields they're obsolete almost immediately in fleet combat.

**Comment by geoffthemedio** [comment]
Use it against another empire's home fleet. Usually survives the battle with other empires' low starting weapon strengths.

There can be various different fighter weapons, including some stronger than 2, which would be more useful against shields. In general, I'd expect fighters to be less useful against shielded opponents than ship weapons, but it depends on the balancing of additional parts.

**Comment by MatGB** [comment]
True, I'm looking at various rock/paper/scissor approaches to ship design, having fighters able to get passed shields before shooting, but always having relatively low damage stats compared to weapons, adds an interesting twist that could reduce the "always use shields" approach most players currently seem to take a bit.

**Comment by MatGB** [comment]
- Number of fighters launched per combat round is determined by fighter bay capacity. Total number of fighters that can be launched depends on hangar capacity.*

This doesn't seem to work correctly, I've got some designs with one hanger, 2 bays, one weapon part but they're only launching 3 fighters each per turn. At 8 fighters in the hanger you want them all launched by round two, round three launches are useless.

Researching a bigger hulltype so I can see if it works with more hangers.

**Comment by geoffthemedio** [comment]
Try now?

**Comment by MatGB** [comment]
Appears to have done it, 6 launched round 1 & 2 on turn 2

File: combat/CombatSystem.cpp [hunk 1]
```diff
@@ -523,24 +720,79 @@ namespace {
...
+            } else if (part_class == PC_FIGHTER_BAY) {
+                part_fighter_launch_capacities[part_name] += ship->CurrentPartMeterValue(METER_CAPACITY, part_name);
+
+            } else if (part_class == PC_FIGHTER_WEAPON) {
+                fighter_attack = std::max(fighter_attack, ship->CurrentPartMeterValue(METER_CAPACITY, part_name));
+            }
+        }
+
+        if (available_fighters > 0 && !part_fighter_launch_capacities.empty()) {
+            for (std::multimap<std::string, int>::iterator launch_it = part_fighter_launch_capacities.begin();
```
Comment (Morlic-fo) [inline combat/CombatSystem.cpp]: Gives a compiler error (MSVC 2010, Win7). Fixed by changing to `std::map::iterator`.

If multimap is indeed intended, need to switch the type of `part_fighter_launch_capacities`...

Comment (geoffthemedio) [inline combat/CombatSystem.cpp]: Do you mean

```
-            for (std::multimap<std::string, int>::iterator launch_it = part_fighter_launch_capacities.begin();
+            for (std::map<std::string, int>::iterator launch_it = part_fighter_launch_capacities.begin();
```

Comment (Morlic-fo) [inline combat/CombatSystem.cpp]: Yes.

**Comment by Morlic-fo** [comment]
Fighter recovery seems not to be working (or works differently than I expect): Eventhough I lose no fighters in combat (all are said to be recovered), no fighters are launched in subsequent combats.

/Edit: #434 seems to fix.

**Comment by Vezzra** [comment]
Looks very interesting and promising. Although I'd prefer some things done differently, the basic idea sounds good.

However, I see that already a bit of design discussions started here, and I don't think this is the place for it. Those discussions belong on the forums. I've [opened a relevant thread](http://freeorion.org/forum/viewtopic.php?f=5&t=9882) on the [Top Priority Game Design](http://freeorion.org/forum/viewforum.php?f=5) forum, please continue all design relevant discussions there.

**Comment by MatGB** [comment]
fighters-ai-0

Picture speaks for itself, the AI's building ships with fighter bays but not hangers.

As a hack it'd be possible to remove fighters from preunlocked and have a cost 1 tech that unlocks it all, IIRC the starting fleets would still exist, right?

**Comment by Morlic-fo** [comment]
It is much easier. In AI/ShipDesignAI.py simply change the line

`WEAPONS = frozenset({fo.shipPartClass.shortRange, fo.shipPartClass.fighterBay})`
to 
`WEAPONS = frozenset({fo.shipPartClass.shortRange})`

**Comment by geoffthemedio** [comment]
Does it matter? The AI needs to be taught to use fighter parts, what are still being designed / implemented. This branch is expected to not yet know how to use the new parts.

**Comment by MatGB** [comment]
It does if you're trying to playtest the parts to see how they work in various circumstances, having the AI wasting resources on unarmed but expensive ships means they're not building warships with, for example, shields, so I can get my head around costings, stats, etc that would be appropriate.

**Comment by Vezzra** [comment]
@geoffthemedio, as you have already (albeit accidentally) merged a rebased version of that branch into master, I think this PR can be closed and the now "derelict" (sort of) Fighters3 branch deleted?

**Pull request closed without merging**